### PR TITLE
Feature nxp 20555 split spreadsheet

### PR DESCRIPTION
**Pull request by adilioaraujo**

### CODE REVIEW

File: nuxeo-platform-spreadsheet-jsf/pom.xml [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+      <artifactId>javax.servlet-api</artifactId>
+    </dependency>
+    <dependency>
+      <groupId>org.nuxeo.ecm.platform</groupId>
+      <artifactId>nuxeo-platform-test</artifactId>
+      <scope>test</scope>
+    </dependency>
+  </dependencies>
+
+  <repositories>
```
Comment (nelsonsilva) [inline nuxeo-platform-spreadsheet-jsf/pom.xml]: Repositories should only be defined in the parent POM

File: nuxeo-platform-spreadsheet-web-ui/src/main/resources/web/nuxeo.war/ui/nuxeo-spreadsheet/nuxeo-spreadsheet-button.html [hunk 2]
```diff
@@ -0,0 +1,139 @@
+<!--
+(C) Copyright 2016 Nuxeo SA (http://nuxeo.com/) and others.
```
Comment (nelsonsilva) [inline nuxeo-platform-spreadsheet-web-ui/src/main/resources/web/nuxeo.war/ui/nuxeo-spreadsheet/nuxeo-spreadsheet-button.html]: 2017 Nuxeo

File: nuxeo-platform-spreadsheet-web-ui/src/main/resources/web/nuxeo.war/ui/nuxeo-spreadsheet/nuxeo-spreadsheet-button.html [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+      },
+
+      _isAvailable: function() {
+        return (this.nxProvider !== null) && this.mode === 'table';
+      },
+
+      _show: function() {
+        // convert from nxProvider.sort to sortInfos
+        var sortInfos = [];
+        Object.keys(this.nxProvider.sort).forEach(function(key) {
```
Comment (nelsonsilva) [inline nuxeo-platform-spreadsheet-web-ui/src/main/resources/web/nuxeo.war/ui/nuxeo-spreadsheet/nuxeo-spreadsheet-button.html]: `var sortInfos =  Object.keys(this.nxProvider.sort).map(...)`

File: nuxeo-platform-spreadsheet-web/src/main/js/app/ui/spreadsheet.js [hunk 4]
```diff
@@ -57,29 +57,95 @@ class Spreadsheet {
...
+          schemasPrefixes.push(schema);
+        }
+      }
+
+      // fetch schemas (based on prefixes)
+      new Schemas(connection).fetch(schemasPrefixes).then((schemas) => {
+
+        // set columns
+        let cols = [];
+        for (let col of resultColumns) {
```
Comment (afjusto) [inline nuxeo-platform-spreadsheet-web/src/main/js/app/ui/spreadsheet.js]: resultColumns.forEach() ?

File: nuxeo-platform-spreadsheet-web-ui/src/main/resources/web/nuxeo.war/ui/nuxeo-spreadsheet/nuxeo-spreadsheet-button.html [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+          sortInfos.push({'sortColumn': key, 'sortAscending': this.nxProvider.sort[key] === 'asc'});
+        }.bind(this));
+
+        // convert nxProvider.params and nxProvider.aggregations to properties
+        var properties = {};
+        Object.keys(this.nxProvider.params).forEach(function(key) {
+          properties[key] = this.nxProvider.params[key];
+        }.bind(this));
```
Comment (nelsonsilva) [inline nuxeo-platform-spreadsheet-web-ui/src/main/resources/web/nuxeo.war/ui/nuxeo-spreadsheet/nuxeo-spreadsheet-button.html]: if you do a `var provider = this.nxProvider;` then you don't have to keep binding all the functions to `this` just to access `this.nxProvider`

File: nuxeo-platform-spreadsheet-web-ui/src/main/resources/web/nuxeo.war/ui/nuxeo-spreadsheet/nuxeo-spreadsheet-button.html [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+
+      _close: function() {
+        this.$.dialog.toggle();
+        this.nxProvider.fetch();
+      },
+
+      _onLoad: function() {
+        var close = (this.$.iframe.contentDocument || this.$.iframe.contentWindow.document).querySelector('#close');
+        if (close) {
+          close.addEventListener('click', function() {
```
Comment (nelsonsilva) [inline nuxeo-platform-spreadsheet-web-ui/src/main/resources/web/nuxeo.war/ui/nuxeo-spreadsheet/nuxeo-spreadsheet-button.html]: `close.addEventListener('click', this._close.bind(this))` should be enough

File: nuxeo-platform-spreadsheet-web-ui/src/main/resources/web/nuxeo.war/ui/nuxeo-spreadsheet/nuxeo-spreadsheet-button.html [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+    </paper-dialog>
+
+  </template>
+
+  <script>
+    Polymer({
+      is: 'nuxeo-spreadsheet-button',
+      behaviors: [Nuxeo.RoutingBehavior, Nuxeo.I18nBehavior, Nuxeo.FiltersBehavior],
+      properties: {
+        baseUrl: {
```
Comment (nelsonsilva) [inline nuxeo-platform-spreadsheet-web-ui/src/main/resources/web/nuxeo.war/ui/nuxeo-spreadsheet/nuxeo-spreadsheet-button.html]: Can´t we get this from a `<nuxeo-connection>`

File: nuxeo-platform-spreadsheet-web-ui/src/main/resources/web/nuxeo.war/ui/nuxeo-spreadsheet/nuxeo-spreadsheet.html [hunk 8]
```diff
@@ -0,0 +1,8 @@
+<link rel="import" href="nuxeo-spreadsheet-button.html">
+
+<nuxeo-slot-content name="spreadsheetButton" slot="DOCUMENT_LISTING_ACTIONS">
+  <template>
+    <nuxeo-spreadsheet-button base-url="[[baseUrl]]" nx-provider="[[nxProvider]]" mode="[[displayMode]]">
```
Comment (nelsonsilva) [inline nuxeo-platform-spreadsheet-web-ui/src/main/resources/web/nuxeo.war/ui/nuxeo-spreadsheet/nuxeo-spreadsheet.html]: Not a fan of the `mode` attribute because the button only works in `table` mode anyway, but I understand the limitations of filtering here without support for expressions.....

File: nuxeo-platform-spreadsheet-web/src/main/js/app/app.js [hunk 9]
```diff
@@ -101,17 +103,17 @@ function run(baseURL = '/nuxeo', username = null, password = null) {
...
-    var pageProvider = (cv) ? cv.pageProviderName : (pp || 'spreadsheet_query');
 
     // Setup the language
-    var language = (nuxeo && nuxeo.spreadsheet && nuxeo.spreadsheet.language) ? nuxeo.spreadsheet.language.split('_')[0] : 'en';
+    let language = (nuxeo && nuxeo.spreadsheet && nuxeo.spreadsheet.language) ? nuxeo.spreadsheet.language.split('_')[0] : 'en';
+
+    // Extract content view configuration
+    let resultLayoutName = (cv && cv.resultLayout && cv.resultLayout.name);
+    let resultColumns = cv && cv.resultColumns;
+    let pageProviderName = (cv) ? cv.pageProviderName : (pp || 'spreadsheet_query');
```
Comment (nelsonsilva) [inline nuxeo-platform-spreadsheet-web/src/main/js/app/app.js]: no need for the parenthesis

File: nuxeo-platform-spreadsheet-web/src/main/js/app/ui/spreadsheet.js [hunk 10]
```diff
@@ -57,29 +57,95 @@ class Spreadsheet {
...
-      var cols = (!columns) ?
-          layout.columns.filter((c) =>  c.selectedByDefault !== false)
-          :
-          columns.map((name) => layout.columns.filter((c) => c.name === name)[0]);
-      this.columns = cols
+    // set columns based on result layout
+    if (resultLayoutName) {
+      new Layout(connection, resultLayoutName, language).fetch().then((layout) => {
+        // Check which columns to display
+        let cols = (!resultColumns) ? layout.columns.filter((c) => c.selectedByDefault !== false)
```
Comment (Gabez0r) [inline nuxeo-platform-spreadsheet-web/src/main/js/app/ui/spreadsheet.js]: Nitpicking: would rather invert the expressions and avoid starting with a negation, for readability.

File: nuxeo-platform-spreadsheet-web/src/main/js/app/ui/spreadsheet.js (hunk 4 above)
Comment (adilioaraujo) [inline nuxeo-platform-spreadsheet-web/src/main/js/app/ui/spreadsheet.js]: Just a matter os preference I guess, and also: 
"With the new for-of loop, ES6 aims to bring the best parts from all three of the previous approaches." https://www.eventbrite.com/engineering/learning-es6-for-of-loop/

File: nuxeo-platform-spreadsheet-web/src/main/js/app/nuxeo/rest/schemas.js [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+class Schemas extends Request {
+
+  constructor(connection) {
+    super(connection, PATH);
+    this.data = {};
+  }
+
+  fetchFieldsBySchema(schema) {
+    return new Promise((resolve, reject) => {
+      this.conn.request(`${PATH}/${schema}?fetch.schema=fields`).repositoryName(undefined).get((error, data) => {
```
Comment (nelsonsilva) [inline nuxeo-platform-spreadsheet-web/src/main/js/app/nuxeo/rest/schemas.js]: I'd rather rely on `return this. execute('get', path).then(...)` or introduce a new method like `fetch` which returned a promise already

File: nuxeo-platform-spreadsheet-web/src/main/js/app/nuxeo/rest/schemas.js [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+        resolve(data);
+      });
+    });
+  }
+
+  fetch(schemas) {
+    this.data = [];
+    return this.execute().then((entries) => {
+
+      for (let entry of entries) {
```
Comment (nelsonsilva) [inline nuxeo-platform-spreadsheet-web/src/main/js/app/nuxeo/rest/schemas.js]: `entries.filter((e) => schemas.indexOf(key) !== -1).forEach((e) => ....)`

File: nuxeo-platform-spreadsheet-jsf/pom.xml [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+    <dependency>
+      <groupId>javax.servlet</groupId>
+      <artifactId>javax.servlet-api</artifactId>
+    </dependency>
+    <dependency>
+      <groupId>org.nuxeo.ecm.platform</groupId>
+      <artifactId>nuxeo-platform-test</artifactId>
+      <scope>test</scope>
+    </dependency>
+  </dependencies>
```
Comment (Gabez0r) [inline nuxeo-platform-spreadsheet-jsf/pom.xml]: Shouldn't it have a dependency on ` nuxeo-platform-spreadsheet-core`?

File: nuxeo-platform-spreadsheet-web/src/main/js/app/nuxeo/rest/schemas.js [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+        let key = entry['@prefix'] || entry.name;
+        if (schemas.indexOf(key) !== -1) {
+          this.data[key] = {name: entry.name};
+        }
+      }
+
+      let promises = [];
```
Comment (nelsonsilva) [inline nuxeo-platform-spreadsheet-web/src/main/js/app/nuxeo/rest/schemas.js]: `const promises = schemas.map((s) => this.fetchFieldsBySchema(this.data[schema].name)`
or just make fetchFieldsBySchema take a schema and make it private if not used elsewhere so you can to `map(this.fetchFieldsBySchema.bind(this)`

File: nuxeo-platform-spreadsheet-web-ui/src/main/resources/OSGI-INF/nuxeo-spreadsheet-webresources-contrib.xml [hunk 15]
```diff
@@ -0,0 +1,21 @@
+<?xml version="1.0"?>
+
+<component name="org.nuxeo.platform.spreadsheet.web.ui.resources.contrib">
+
+  <require>org.nuxeo.web.ui.resources</require>
+
+  <extension target="org.nuxeo.ecm.platform.WebResources" point="resources">
+    <resource name="nuxeo-spreadsheet.html" type="import" shrinkable="false">
```
Comment (Gabez0r) [inline nuxeo-platform-spreadsheet-web-ui/src/main/resources/OSGI-INF/nuxeo-spreadsheet-webresources-contrib.xml]: Do we really need the `shrinkable="false"` here?

File: nuxeo-platform-spreadsheet-web/src/main/js/app/nuxeo/rest/schemas.js [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+      for (let schema of schemas) {
+        promises.push(this.fetchFieldsBySchema(this.data[schema].name));
+      }
+
+      return Promise.all(promises).then(values => {
+        return this.data;
+      }, reason => {
+        console.log(reason);
```
Comment (nelsonsilva) [inline nuxeo-platform-spreadsheet-web/src/main/js/app/nuxeo/rest/schemas.js]: would rather not catch the error just to log it to the console....

File: nuxeo-platform-spreadsheet-web/src/main/js/app/ui/spreadsheet.js [hunk 17]
```diff
@@ -24,10 +25,9 @@ import {Query} from '../nuxeo/rpc/query';
  */
 class Spreadsheet {
 
-  constructor(container, connection, layout, columns, pageProvider, language) {
+  constructor(container, connection, resultLayoutName, resultColumns, pageProviderName, language) {
```
Comment (nelsonsilva) [inline nuxeo-platform-spreadsheet-web/src/main/js/app/ui/spreadsheet.js]: why the change of names? these are just longer and in this context there's no mix up, we have one layout, one set of columns and one pageProvider....

File: nuxeo-platform-spreadsheet-jsf/pom.xml (hunk 13 above)
Comment (Gabez0r) [inline nuxeo-platform-spreadsheet-jsf/pom.xml]: And ` nuxeo-platform-spreadsheet-web`?

File: nuxeo-platform-spreadsheet-web/src/main/js/app/ui/spreadsheet.js [hunk 18]
```diff
@@ -57,29 +57,95 @@ class Spreadsheet {
...
-
-    new Layout(connection, layout, language).fetch().then((layout) => {
-      // Check which columns to display
-      var cols = (!columns) ?
-          layout.columns.filter((c) =>  c.selectedByDefault !== false)
-          :
-          columns.map((name) => layout.columns.filter((c) => c.name === name)[0]);
-      this.columns = cols
+    // set columns based on result layout
+    if (resultLayoutName) {
```
Comment (nelsonsilva) [inline nuxeo-platform-spreadsheet-web/src/main/js/app/ui/spreadsheet.js]: I would have preferred if we handled these two scenarios in the app and the spreadsheet just got the columns as constructor argument....

File: nuxeo-platform-spreadsheet-web/src/main/js/app/ui/spreadsheet.js [hunk 19]
```diff
@@ -57,29 +57,95 @@ class Spreadsheet {
...
           // Only show columns with a known widget type and with a field
           .filter((c) => c.hasSupportedWidgetType && c.field);
-    });
+      });
+
+    // or based on result columns only
+    } else {
+
+      // default columns
+      if (!resultColumns) {
```
Comment (nelsonsilva) [inline nuxeo-platform-spreadsheet-web/src/main/js/app/ui/spreadsheet.js]: these default columns would be handled in the app too... ideally in the future people would be able to configure these

File: nuxeo-platform-spreadsheet-web/src/main/js/app/ui/spreadsheet.js [hunk 20]
```diff
@@ -57,29 +57,95 @@ class Spreadsheet {
...
+        resultColumns = [
+          { label: 'Title', field: 'dc:title' },
+          { label: 'Modified', field: 'dc:modified'},
+          { label: 'Last Contributor', field: 'dc:lastContributor'},
+          { label: 'State', field: 'currentLifeCycleState'}
+        ];
+      }
+
+      // get schemas prefixes from columns
+      let schemasPrefixes = [];
```
Comment (nelsonsilva) [inline nuxeo-platform-spreadsheet-web/src/main/js/app/ui/spreadsheet.js]: `const schemasPrefixes = resultColumns.filter(....).map(....)`

File: nuxeo-platform-spreadsheet-web-ui/pom.xml [hunk 21]
```diff
@@ -0,0 +1,15 @@
...
+  <parent>
+    <groupId>org.nuxeo.ecm.platform</groupId>
+    <artifactId>nuxeo-platform-spreadsheet</artifactId>
+    <version>8.10-SNAPSHOT</version>
+    <relativePath>../pom.xml</relativePath>
+  </parent>
+
+  <artifactId>nuxeo-platform-spreadsheet-web-ui</artifactId>
+  <name>Nuxeo Platform Spreadsheet Web UI</name>
+  <description>Nuxeo Platform Spreadsheet Web UI</description>
```
Comment (Gabez0r) [inline nuxeo-platform-spreadsheet-web-ui/pom.xml]: No dependencies? Doesn't it depend on `nuxeo-platform-spreadsheet-core` and  `nuxeo-platform-spreadsheet-web` at least?

File: nuxeo-platform-spreadsheet-web/src/main/js/app/ui/spreadsheet.js [hunk 22]
```diff
@@ -57,29 +57,95 @@ class Spreadsheet {
...
+        if (schema && schemasPrefixes.indexOf(schema) === -1) {
+          schemasPrefixes.push(schema);
+        }
+      }
+
+      // fetch schemas (based on prefixes)
+      new Schemas(connection).fetch(schemasPrefixes).then((schemas) => {
+
+        // set columns
+        let cols = [];
```
Comment (nelsonsilva) [inline nuxeo-platform-spreadsheet-web/src/main/js/app/ui/spreadsheet.js]: `const cols = resultColumns.map(....)`

File: nuxeo-platform-spreadsheet-web/src/main/js/app/ui/spreadsheet.js [hunk 23]
```diff
@@ -57,29 +57,95 @@ class Spreadsheet {
...
+        for (let col of resultColumns) {
+
+          let column = {};
```
Comment (nelsonsilva) [inline nuxeo-platform-spreadsheet-web/src/main/js/app/ui/spreadsheet.js]: `let column = { def: ...., widget: ...}`

File: nuxeo-platform-spreadsheet-web/src/main/js/app/ui/spreadsheet.js [hunk 24]
```diff
@@ -57,29 +57,95 @@ class Spreadsheet {
...
+          column.def = {properties: {any: {sortPropertyName: col.field, label: col.label}}};
+          column.widget = {field: col.field};
+
+          // get field from schemas map
+          let field = undefined; // <- explicitly set field as undefined in each iteration
```
Comment (nelsonsilva) [inline nuxeo-platform-spreadsheet-web/src/main/js/app/ui/spreadsheet.js]: why do we need to set it to `undefined`?

File: nuxeo-platform-spreadsheet-web/src/main/js/app/ui/spreadsheet.js [hunk 25]
```diff
@@ -57,29 +57,95 @@ class Spreadsheet {
...
+          if (col.field.includes(':')) {
+            let s = col.field.split(':')[0];
```
Comment (nelsonsilva) [inline nuxeo-platform-spreadsheet-web/src/main/js/app/ui/spreadsheet.js]: `let [s, j] = col.field.split(':')`

File: nuxeo-platform-spreadsheet-web/src/main/js/app/ui/spreadsheet.js [hunk 26]
```diff
@@ -57,29 +57,95 @@ class Spreadsheet {
...
+            let f = col.field.split(':')[1];
+            field = schemas[s].fields[f] || undefined;
```
Comment (nelsonsilva) [inline nuxeo-platform-spreadsheet-web/src/main/js/app/ui/spreadsheet.js]: why do we need the `|| undefined` ?

File: nuxeo-platform-spreadsheet-web/src/main/js/app/ui/spreadsheet.js [hunk 27]
```diff
@@ -57,29 +57,95 @@ class Spreadsheet {
...
+          }
+
+          // set column widget type and properties based on field constraints
+          if (field) {
```
Comment (nelsonsilva) [inline nuxeo-platform-spreadsheet-web/src/main/js/app/ui/spreadsheet.js]: isn't field only defined when you enter the previous if block ? if so then I guess this should be in that same block... or just do continue  if (!col.field.includes(':'))

File: nuxeo-platform-spreadsheet-web/src/main/js/app/ui/spreadsheet.js (hunk 24 above)
Comment (adilioaraujo) [inline nuxeo-platform-spreadsheet-web/src/main/js/app/ui/spreadsheet.js]: Weird bug that took some time to track down... if didn't explicitly set undefined, the field kept the value from last iteration?? Maybe I'm missing something, but this solved it.

File: nuxeo-platform-spreadsheet-jsf/pom.xml (hunk 13 above)
Comment (adilioaraujo) [inline nuxeo-platform-spreadsheet-jsf/pom.xml]: Done

File: nuxeo-platform-spreadsheet-web-ui/pom.xml (hunk 21 above)
Comment (adilioaraujo) [inline nuxeo-platform-spreadsheet-web-ui/pom.xml]: Done

File: nuxeo-platform-spreadsheet-web/src/main/js/app/ui/spreadsheet.js (hunk 26 above)
Comment (adilioaraujo) [inline nuxeo-platform-spreadsheet-web/src/main/js/app/ui/spreadsheet.js]: Because s can be undefined

File: nuxeo-platform-spreadsheet-web/src/main/js/app/ui/spreadsheet.js (hunk 17 above)
Comment (adilioaraujo) [inline nuxeo-platform-spreadsheet-web/src/main/js/app/ui/spreadsheet.js]: IMHO the names better represent the values they are holding, and it confused because we already have this.columns, and layout vars in the same scope. But I've made the changes anyway.

File: nuxeo-platform-spreadsheet-web/src/main/js/app/ui/spreadsheet.js (hunk 18 above)
Comment (adilioaraujo) [inline nuxeo-platform-spreadsheet-web/src/main/js/app/ui/spreadsheet.js]: I know what you mean... needs some refactoring thought, do you consider it a priority?

**Comment by nelsonsilva** [comment]
👍

**Comment by Gabez0r** [comment]
👍
